stl/inc/ranges

```diff
@@ -33,10 +33,195 @@ namespace ranges {
+
     template <class _Ty>
     concept _Copy_constructible_object = copy_constructible<_Ty> && is_object_v<_Ty>;
     // clang-format on
 
+    namespace _Pipe {
+        template <class>
+        struct _Base;
+
+        template <class, class>
+        struct _Pipeline;
+
+        // clang-format off
+        template <class _Left, class _Right>
+        concept _Can_pipe = requires(_Left&& __l, _Right&& __r) {
+            static_cast<_Right&&>(__r)(static_cast<_Left&&>(__l));
+        };
+
+        template <class _Left, class _Right>
+        concept _Can_compose = constructible_from<remove_cvref_t<_Left>, _Left>
+            && constructible_from<remove_cvref_t<_Right>, _Right>;
+        // clang-format on
+
+        template <class _Derived>
+        struct _Base {
+            // clang-format off
+            template <class _Other>
+                requires _Can_compose<_Derived&&, _Other&&>
+            friend constexpr auto operator|(_Base&& __l, _Base<_Other>&& __r) noexcept(noexcept(
+                _Pipeline{static_cast<_Derived&&>(__l), static_cast<_Other&&>(__r)})) {
+                // clang-format on
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Derived, _Base<_Derived>>);
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Other, _Base<_Other>>);
+                return _Pipeline{static_cast<_Derived&&>(__l), static_cast<_Other&&>(__r)};
+            }
+
+            // clang-format off
+            template <class _Other>
+                requires _Can_compose<_Derived&&, const _Other&>
+            friend constexpr auto operator|(_Base&& __l, const _Base<_Other>& __r) noexcept(noexcept(
+                _Pipeline{static_cast<_Derived&&>(__l), static_cast<const _Other&>(__r)})) {
+                // clang-format on
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Derived, _Base<_Derived>>);
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Other, _Base<_Other>>);
+                return _Pipeline{static_cast<_Derived&&>(__l), static_cast<const _Other&>(__r)};
+            }
+
+            // clang-format off
+            template <class _Other>
+                requires _Can_compose<const _Derived&, _Other&&>
+            friend constexpr auto operator|(const _Base& __l, _Base<_Other>&& __r) noexcept(noexcept(
+                _Pipeline{static_cast<const _Derived&>(__l), static_cast<_Other&&>(__r)})) {
+                // clang-format on
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Derived, _Base<_Derived>>);
+                _STL_INTERNAL_STATIC_ASSERT(derived_from<_Other, _Base<_Other>>);
+                return _Pipeline{static_cast<const _Derived&>(__l), static_cast<_Other&&>(__r)};
+            }
+
+            // clang-format off
+            template <class _Other>
+                requires _Can_compose<const _Derived&, const _Other&>
+            friend constexpr auto operator|(const _Base& __l, const _Base<_Other>& __r) noexcept(noexcept(

```

Comment:
Is there no way to avoid overloading this 4x for all combinations of lvalues/rvalues? I assume something is happening with concepts here that prevents us from just taking everything through perfect forwarding and using metaprogramming to determine the derived class, but I don't immediately see why it would be impossible.

Reply:
I believe we have to wait for the "deducing this" paper to simplify this

Reply:
I could simplify these a bit by making them members, but as @miscco implies, we need to deduce the type and value category of `*this` to perfect-forward it into the `_Pipeline` construction.